// create-history/lib/__tests__/describePush.js
'use strict';

exports.__esModule = true;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _expect = require('expect');

var _expect2 = _interopRequireDefault(_expect);

var _Actions = require('../Actions');

var _execSteps = require('./execSteps');

var _execSteps2 = _interopRequireDefault(_execSteps);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var describePush = function describePush(createHistory) {
  describe('push', function () {
    var history = void 0;
    beforeEach(function () {
      history = createHistory();
    });

    describe('with a path string', function () {
      it('calls change listeners with the new location', function (done) {
        var steps = [function (location) {
          (0, _expect2.default)(location.pathname).toEqual('/');
          (0, _expect2.default)(location.search).toEqual('');
          (0, _expect2.default)(location.state).toBe(undefined);
          (0, _expect2.default)(location.action).toEqual(_Actions.POP);
          (0, _expect2.default)(location.key).toBe(null);

          history.push('/home?the=query');
        }, function (location) {
          (0, _expect2.default)(location.pathname).toEqual('/home');
          (0, _expect2.default)(location.search).toEqual('?the=query');
          (0, _expect2.default)(location.state).toBe(undefined);
          (0, _expect2.default)(location.action).toEqual(_Actions.PUSH);
          (0, _expect2.default)(location.key).toExist();
        }];

        (0, _execSteps2.default)(steps, history, done);
      });

      it('should trigger only once when path is changed', function (done) {
        var spy = _expect2.default.createSpy();
        var unlisten = history.listen(spy);
        history.push('/test');
        setTimeout(function () {
          (0, _expect2.default)(spy.calls.length).toBe(1);
          unlisten();
          done();
        }, 10);
      });
    });

    describe('with a path object', function () {
      it('calls change listeners with the new location', function (done) {
        var steps = [function (location) {
          (0, _expect2.default)(location.pathname).toEqual('/');
          (0, _expect2.default)(location.search).toEqual('');
          (0, _expect2.default)(location.state).toBe(undefined);
          (0, _expect2.default)(location.action).toEqual(_Actions.POP);
          (0, _expect2.default)(location.key).toBe(null);

          history.push({
            pathname: '/home',
            search: '?the=query',
            state: { the: 'state' }
          });
        }, function (location) {
          (0, _expect2.default)(location.pathname).toEqual('/home');
          (0, _expect2.default)(location.search).toEqual('?the=query');
          (0, _expect2.default)(location.state).toEqual({ the: 'state' });
          (0, _expect2.default)(location.action).toEqual(_Actions.PUSH);
          (0, _expect2.default)(location.key).toExist();
        }];

        (0, _execSteps2.default)(steps, history, done);
      });

      it('correctly merges with old location', function (done) {
        var oldLocation = void 0;

        var steps = [function (location) {
          (0, _expect2.default)(location.pathname).toEqual('/');
          (0, _expect2.default)(location.search).toEqual('');
          (0, _expect2.default)(location.state).toBe(undefined);
          (0, _expect2.default)(location.action).toEqual(_Actions.POP);
          (0, _expect2.default)(location.key).toBe(null);

          oldLocation = location;

          history.push(_extends({}, location, {
            search: '?the=query',
            state: { the: 'state' }
          }));
        }, function (location) {
          (0, _expect2.default)(location.pathname).toEqual(oldLocation.pathname);
          (0, _expect2.default)(location.search).toEqual('?the=query');
          (0, _expect2.default)(location.state).toEqual({ the: 'state' });
          (0, _expect2.default)(location.action).toEqual(_Actions.PUSH);
          (0, _expect2.default)(location.key).toExist();
          (0, _expect2.default)(location.key).toNotEqual(oldLocation.key);
        }];

        (0, _execSteps2.default)(steps, history, done);
      });

      it('becomes a REPLACE if path is unchanged', function (done) {
        var steps = [function (location) {
          (0, _expect2.default)(location.pathname).toEqual('/');
          (0, _expect2.default)(location.search).toEqual('');
          (0, _expect2.default)(location.state).toBe(undefined);
          (0, _expect2.default)(location.action).toEqual(_Actions.POP);
          (0, _expect2.default)(location.key).toBe(null);

          history.push({
            pathname: '/home',
            search: '?the=query',
            state: { the: 'state' }
          });
        }, function (location) {
          (0, _expect2.default)(location.pathname).toEqual('/home');
          (0, _expect2.default)(location.search).toEqual('?the=query');
          (0, _expect2.default)(location.state).toEqual({ the: 'state' });
          (0, _expect2.default)(location.action).toEqual(_Actions.PUSH);
          (0, _expect2.default)(location.key).toExist();

          history.push({
            pathname: '/home',
            search: '?the=query',
            state: { the: 'state' }
          });
        }, function (location) {
          (0, _expect2.default)(location.pathname).toEqual('/home');
          (0, _expect2.default)(location.search).toEqual('?the=query');
          (0, _expect2.default)(location.state).toEqual({ the: 'state' });
          (0, _expect2.default)(location.action).toEqual(_Actions.REPLACE);
          (0, _expect2.default)(location.key).toExist();
        }];

        (0, _execSteps2.default)(steps, history, done);
      });

      it('stays PUSH if state is changed', function (done) {
        var steps = [function (location) {
          (0, _expect2.default)(location.pathname).toEqual('/');
          (0, _expect2.default)(location.search).toEqual('');
          (0, _expect2.default)(location.state).toBe(undefined);
          (0, _expect2.default)(location.action).toEqual(_Actions.POP);
          (0, _expect2.default)(location.key).toBe(null);

          history.push({
            pathname: '/home',
            search: '?the=query',
            state: { the: 'state' }
          });
        }, function (location) {
          (0, _expect2.default)(location.pathname).toEqual('/home');
          (0, _expect2.default)(location.search).toEqual('?the=query');
          (0, _expect2.default)(location.state).toEqual({ the: 'state' });
          (0, _expect2.default)(location.action).toEqual(_Actions.PUSH);
          (0, _expect2.default)(location.key).toExist();

          history.push({
            pathname: '/home',
            search: '?the=query',
            state: { different: 'state' }
          });
        }, function (location) {
          (0, _expect2.default)(location.pathname).toEqual('/home');
          (0, _expect2.default)(location.search).toEqual('?the=query');
          (0, _expect2.default)(location.state).toEqual({ different: 'state' });
          (0, _expect2.default)(location.action).toEqual(_Actions.PUSH);
          (0, _expect2.default)(location.key).toExist();
        }];

        (0, _execSteps2.default)(steps, history, done);
      });
    });
  });
};

exports.default = describePush;
