// create-history/lib/__tests__/describeReplace.js
'use strict';

exports.__esModule = true;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _expect = require('expect');

var _expect2 = _interopRequireDefault(_expect);

var _Actions = require('../Actions');

var _execSteps = require('./execSteps');

var _execSteps2 = _interopRequireDefault(_execSteps);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var describeReplace = function describeReplace(createHistory) {
  describe('replace', function () {
    var history = void 0;
    beforeEach(function () {
      history = createHistory();
    });

    describe('with a path string', function () {
      it('calls change listeners with the new location', function (done) {
        var steps = [function (location) {
          (0, _expect2.default)(location.pathname).toEqual('/');
          (0, _expect2.default)(location.search).toEqual('');
          (0, _expect2.default)(location.state).toBe(undefined);
          (0, _expect2.default)(location.action).toEqual(_Actions.POP);

          history.replace('/home?the=query');
        }, function (location) {
          (0, _expect2.default)(location.pathname).toEqual('/home');
          (0, _expect2.default)(location.search).toEqual('?the=query');
          (0, _expect2.default)(location.state).toBe(undefined);
          (0, _expect2.default)(location.action).toEqual(_Actions.REPLACE);
        }];

        (0, _execSteps2.default)(steps, history, done);
      });
    });

    describe('with a path object', function () {
      it('calls change listeners with the new location', function (done) {
        var steps = [function (location) {
          (0, _expect2.default)(location.pathname).toEqual('/');
          (0, _expect2.default)(location.search).toEqual('');
          (0, _expect2.default)(location.state).toBe(undefined);
          (0, _expect2.default)(location.action).toEqual(_Actions.POP);

          history.replace({
            pathname: '/home',
            search: '?the=query',
            state: { the: 'state' }
          });
        }, function (location) {
          (0, _expect2.default)(location.pathname).toEqual('/home');
          (0, _expect2.default)(location.search).toEqual('?the=query');
          (0, _expect2.default)(location.state).toEqual({ the: 'state' });
          (0, _expect2.default)(location.action).toEqual(_Actions.REPLACE);
        }];

        (0, _execSteps2.default)(steps, history, done);
      });

      it('correctly merges with old location', function (done) {
        var oldLocation = void 0;

        var steps = [function (location) {
          (0, _expect2.default)(location.pathname).toEqual('/');
          (0, _expect2.default)(location.search).toEqual('');
          (0, _expect2.default)(location.state).toBe(undefined);
          (0, _expect2.default)(location.action).toEqual(_Actions.POP);

          oldLocation = location;

          history.replace(_extends({}, location, {
            search: '?the=query',
            state: { the: 'state' }
          }));
        }, function (location) {
          (0, _expect2.default)(location.pathname).toEqual(oldLocation.pathname);
          (0, _expect2.default)(location.search).toEqual('?the=query');
          (0, _expect2.default)(location.state).toEqual({ the: 'state' });
          (0, _expect2.default)(location.action).toEqual(_Actions.REPLACE);
          (0, _expect2.default)(location.key).toNotEqual(oldLocation.key);
        }];

        (0, _execSteps2.default)(steps, history, done);
      });
    });
  });
};

exports.default = describeReplace;
