// create-history/lib/__tests__/describeHashSupport.js
'use strict';

exports.__esModule = true;

var _expect = require('expect');

var _expect2 = _interopRequireDefault(_expect);

var _Actions = require('../Actions');

var _execSteps = require('./execSteps');

var _execSteps2 = _interopRequireDefault(_execSteps);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var describeHashSupport = function describeHashSupport(createHistory) {
  describe('when a URL with a hash is pushed', function () {
    var history = void 0;
    beforeEach(function () {
      history = createHistory();
    });

    it('preserves the hash', function (done) {
      var steps = [function (location) {
        (0, _expect2.default)(location.pathname).toEqual('/');
        (0, _expect2.default)(location.search).toEqual('');
        (0, _expect2.default)(location.hash).toEqual('');
        (0, _expect2.default)(location.state).toBe(undefined);
        (0, _expect2.default)(location.action).toEqual(_Actions.POP);
        (0, _expect2.default)(location.key).toBe(null);

        history.push({
          pathname: '/home',
          search: '?the=query',
          hash: '#the-hash',
          state: { the: 'state' }
        });
      }, function (location) {
        (0, _expect2.default)(location.pathname).toEqual('/home');
        (0, _expect2.default)(location.search).toEqual('?the=query');
        (0, _expect2.default)(location.hash).toEqual('#the-hash');
        (0, _expect2.default)(location.state).toEqual({ the: 'state' });
        (0, _expect2.default)(location.action).toEqual(_Actions.PUSH);
        (0, _expect2.default)(location.key).toExist();
      }];

      (0, _execSteps2.default)(steps, history, done);
    });

    it('does not convert PUSH to REPLACE if path does not change', function (done) {
      var steps = [function (location) {
        (0, _expect2.default)(location.pathname).toEqual('/');
        (0, _expect2.default)(location.search).toEqual('');
        (0, _expect2.default)(location.hash).toEqual('');
        (0, _expect2.default)(location.state).toBe(undefined);
        (0, _expect2.default)(location.action).toEqual(_Actions.POP);
        (0, _expect2.default)(location.key).toBe(null);

        history.push('/#the-hash');
      }, function (location) {
        (0, _expect2.default)(location.pathname).toEqual('/');
        (0, _expect2.default)(location.search).toEqual('');
        (0, _expect2.default)(location.hash).toEqual('#the-hash');
        (0, _expect2.default)(location.state).toBe(undefined);
        (0, _expect2.default)(location.action).toEqual(_Actions.PUSH);
        (0, _expect2.default)(location.key).toExist();
      }];

      (0, _execSteps2.default)(steps, history, done);
    });
  });
};

exports.default = describeHashSupport;
