// create-history/lib/__tests__/describeTransitions.js
'use strict';

exports.__esModule = true;

var _expect = require('expect');

var _expect2 = _interopRequireDefault(_expect);

var _Actions = require('../Actions');

var _execSteps = require('./execSteps');

var _execSteps2 = _interopRequireDefault(_execSteps);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var describeTransitions = function describeTransitions(createHistory) {
  describe('a synchronous transition hook', function () {
    var history = void 0,
        unlistenBefore = void 0;
    beforeEach(function () {
      history = createHistory();
    });

    afterEach(function () {
      if (unlistenBefore) unlistenBefore();
    });

    it('receives the next location', function (done) {
      var nextLocation = void 0;
      var steps = [function () {
        history.push({
          pathname: '/home',
          search: '?the=query',
          state: { the: 'state' }
        });
      }, function (location) {
        (0, _expect2.default)(nextLocation).toBe(location);
      }];

      unlistenBefore = history.listenBefore(function (location) {
        nextLocation = location;
      });

      (0, _execSteps2.default)(steps, history, done);
    });
  });

  describe('an asynchronous transition hook', function () {
    var history = void 0,
        unlistenBefore = void 0;
    beforeEach(function () {
      history = createHistory();
    });

    afterEach(function () {
      if (unlistenBefore) unlistenBefore();
    });

    it('receives the next location', function (done) {
      var nextLocation = void 0;
      var steps = [function () {
        history.push({
          pathname: '/home',
          search: '?the=query',
          state: { the: 'state' }
        });
      }, function (location) {
        (0, _expect2.default)(nextLocation).toBe(location);
      }];

      unlistenBefore = history.listenBefore(function (location, callback) {
        nextLocation = location;
        setTimeout(callback);
      });

      (0, _execSteps2.default)(steps, history, done);
    });
  });

  describe('when the user confirms a transition', function () {
    var location = void 0,
        history = void 0,
        unlisten = void 0,
        unlistenBefore = void 0;
    beforeEach(function () {
      location = null;

      var confirmationMessage = 'Are you sure?';

      history = createHistory({
        getUserConfirmation: function getUserConfirmation(message, callback) {
          (0, _expect2.default)(message).toBe(confirmationMessage);
          callback(true);
        }
      });

      unlistenBefore = history.listenBefore(function () {
        return confirmationMessage;
      });

      unlisten = history.listen(function (loc) {
        location = loc;
      });
    });

    afterEach(function () {
      if (unlistenBefore) unlistenBefore();

      if (unlisten) unlisten();
    });

    it('updates the location', function () {
      var prevLocation = location;

      history.push({
        pathname: '/home',
        search: '?the=query',
        state: { the: 'state' }
      });

      (0, _expect2.default)(prevLocation).toNotBe(location);

      (0, _expect2.default)(location.pathname).toEqual('/home');
      (0, _expect2.default)(location.search).toEqual('?the=query');
      (0, _expect2.default)(location.state).toEqual({ the: 'state' });
      (0, _expect2.default)(location.action).toEqual(_Actions.PUSH);
      (0, _expect2.default)(location.key).toExist();
    });
  });

  describe('when the user cancels a transition', function () {
    var location = void 0,
        history = void 0,
        unlisten = void 0,
        unlistenBefore = void 0;
    beforeEach(function () {
      location = null;

      var confirmationMessage = 'Are you sure?';

      history = createHistory({
        getUserConfirmation: function getUserConfirmation(message, callback) {
          (0, _expect2.default)(message).toBe(confirmationMessage);
          callback(false);
        }
      });

      unlistenBefore = history.listenBefore(function () {
        return confirmationMessage;
      });

      unlisten = history.listen(function (loc) {
        location = loc;
      });
    });

    afterEach(function () {
      if (unlistenBefore) unlistenBefore();

      if (unlisten) unlisten();
    });

    it('does not update the location', function () {
      var prevLocation = location;
      history.push('/home');
      (0, _expect2.default)(prevLocation).toBe(location);
    });
  });

  describe('when the transition hook cancels a transition', function () {
    var location = void 0,
        history = void 0,
        unlisten = void 0,
        unlistenBefore = void 0;
    beforeEach(function () {
      location = null;

      history = createHistory();

      unlistenBefore = history.listenBefore(function () {
        return false;
      });

      unlisten = history.listen(function (loc) {
        location = loc;
      });
    });

    afterEach(function () {
      if (unlistenBefore) unlistenBefore();

      if (unlisten) unlisten();
    });

    it('does not update the location', function () {
      var prevLocation = location;
      history.push('/home');
      (0, _expect2.default)(prevLocation).toBe(location);
    });
  });
};

exports.default = describeTransitions;
