// create-history/lib/__tests__/describePopStateCancel.js
'use strict';

exports.__esModule = true;

var _expect = require('expect');

var _expect2 = _interopRequireDefault(_expect);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var describePopStateCancel = function describePopStateCancel(createHistory) {
  describe('when popstate transitons are canceled', function () {
    var history = void 0,
        unlistenBefore = void 0;
    beforeEach(function () {
      history = createHistory();
      history.push('/a');
      history.push('/b');
      history.push('/c');

      unlistenBefore = history.listenBefore(function () {
        return false;
      });
    });

    afterEach(function () {
      if (unlistenBefore) unlistenBefore();
    });

    it('restores the previous location', function (done) {
      window.history.back();

      setTimeout(function () {
        var currentLocation = history.getCurrentLocation();
        (0, _expect2.default)(currentLocation.pathname).toBe('/c');
        done();
      }, 100);
    });
  });
};

exports.default = describePopStateCancel;
